### ISSUE TITLE
Knex returns Result object instead of empty array of results

### ISSUE THREAD

**Issue by sdemjanenko** [knex/knex#4425]
I've noticed the following behavior
```
.insert([]).returning(["id"]) // Returns Result
.insert([{...something}]).returning(["id"]) // returns [{id:...}]
.insert([{...something with conflict}]).returning(["id"]).onConflict().ignore() // returns []
```

The behavior of the 2nd and 3rd cases is what I expect. I do not expect a Result object to ever be returned when I provide a `returning` as part of the query.

**Comment by elhigu** [comment]
shouldn't first one throw an exception from database because of invalid query? also not following issue template and missing information.

**Comment by elhigu** [comment]
yeah... with pg it generates completely empty query https://runkit.com/embed/5ykovnc5ix0v might run with success, but I would rather say that it is invalid usage of knex.

**Comment by sdemjanenko** [comment]
Given that insertions may be stored in a variable and that variable may be an empty array, I think knex should do the expected thing and return the same type (or throw an error). Returning a Result object certainly is not what I expected it to do and led to a runtime error in my code.

**Comment by mercmobily** [comment]
This likely needs clarification and possibly a guard. Passing an empty array to `.insert([])` is effectively an empty operation; behaviour varies per dialect/driver, and on PostgreSQL the internal path can currently surface a driver result wrapper instead of an empty array when `returning()` is present. It’s not a supported shape to begin with.

What Knex compilers do today for empty inserts:

- The generic compiler returns an empty SQL string for `insert([])` (see `lib/query/querycompiler.js:205–244`), meaning nothing is executed.
- With `returning()`, PostgreSQL would usually return an array of rows for non‑empty inserts, but for an empty insert there are no rows to return. Returning a “Result” wrapper in this case is unhelpful.

Two ideas we could consider:

- Treat `insert([])` as invalid usage and throw a clear error before execution, regardless of `returning`.
- Or normalise the result to an empty array for `returning()` calls with an empty payload for consistency.

Given that the first option fails fast on a likely logic bug (silently doing nothing), it would be my preference. If your application legitimately constructs empty batches at times, guarding the call before passing it to Knex is safest.

This needs to be decided --- we will do so once the mass triaging (nearly finished) is done and we get to this issue. I am filing it as a feature though.
